src/Microsoft.AspNetCore.Server.Kestrel/Internal/Infrastructure/MemoryPoolIteratorExtensions.cs

```diff
@@ -158,6 +164,37 @@ public static bool GetKnownMethod(this MemoryPoolIterator begin, out string know
         }
 
         /// <summary>
+        /// Checks 8 bytes from <paramref name="begin"/> that correspond to 'http://' or 'https://'
+        /// </summary>
+        /// <param name="begin">The iterator</param>
+        /// <param name="knownScheme">A reference to the known scheme, if the input matches any</param>
+        /// <returns>True when memory starts with known http or https schema</returns>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static bool GetKnownHttpSchema(this MemoryPoolIterator begin, out string knownScheme)
+        {
+            knownScheme = null;
+            ulong value;
+            if (!begin.TryPeekLong(out value))
+            {
+                return false;
+            }
+
+            if ((value & _mask7Chars) == _httpSchemeLong)
```

Comment:
Technically it's not a valid client scenario to send absolute-form targets to the end-server, right? It's only meant to be sent to proxies AFAIK. I recall someone mentioning the spec requires that proxies change the target to origin-form, but end-servers still support absolute-form in case the proxy isn't compliant with the spec.

If we already going through this work to support buggy proxies, I think we should also allow ws/wss.